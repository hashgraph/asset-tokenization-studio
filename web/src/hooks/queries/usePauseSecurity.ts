import { UseMutationOptions, useMutation } from "@tanstack/react-query";
import { SDKService } from "../../services/SDKService";
import { PauseRequest } from "@iob/securitytoken-sdk";
import { useToast } from "@iob/io-bricks-ui/Overlay/Toast";
import { useTranslation } from "react-i18next";

export const usePauseSecurity = (
  options: UseMutationOptions<boolean, unknown, PauseRequest, unknown> = {},
) => {
  const toast = useToast();
  const { t } = useTranslation("security", { keyPrefix: "pause" });

  return useMutation(
    (pauseRequest: PauseRequest) => SDKService.pause(pauseRequest),
    {
      onSuccess: (data) => {
        if (data === true) {
          toast.show({
            duration: 3000,
            title: `${t("messages.success")}`,
            description: `${t("messages.descriptionSuccess")}`,
            variant: "subtle",
            status: "success",
          });
        } else {
          toast.show({
            duration: 3000,
            title: t("messages.error"),
            description: t("messages.descriptionFailed"),
            variant: "subtle",
            status: "error",
          });
        }
      },
      onError: (error) => {
        console.log("SDK message --> Security pause error: ", error);
      },
      ...options,
    },
  );
};
