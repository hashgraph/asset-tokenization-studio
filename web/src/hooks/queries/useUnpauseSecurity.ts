import { UseMutationOptions, useMutation } from "@tanstack/react-query";
import { SDKService } from "../../services/SDKService";
import { PauseRequest } from "@hashgraph/securitytoken-sdk";
import { useToast } from "@hashgraph/securitytoken-uicomponents/Overlay/Toast";
import { useTranslation } from "react-i18next";

export const useUnpauseSecurity = (
  options: UseMutationOptions<boolean, unknown, PauseRequest, unknown> = {},
) => {
  const toast = useToast();
  const { t } = useTranslation("security", { keyPrefix: "unpause" });

  return useMutation(
    (pauseRequest: PauseRequest) => SDKService.unpause(pauseRequest),
    {
      onSuccess: (data) => {
        if (data === true) {
          toast.show({
            duration: 3000,
            title: `${t("messages.success")}`,
            description: `${t("messages.descriptionSuccess")}`,
            variant: "subtle",
            status: "success",
          });
        } else {
          toast.show({
            duration: 3000,
            title: t("messages.error"),
            description: t("messages.descriptionFailed"),
            variant: "subtle",
            status: "error",
          });
        }
      },
      onError: (error) => {
        console.log("SDK message --> Security pause error: ", error);
      },
      ...options,
    },
  );
};
