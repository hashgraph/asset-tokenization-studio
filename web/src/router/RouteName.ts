export enum RouteName {
  Landing = "LANDING",
  Dashboard = "DASHBOARD",
  AddSecurity = "ADD_SECURITY",
  CreateSecurity = "CREATE_SECURITY",
  CreateEquity = "CREATE_EQUITY",
  CreateBond = "CREATE_BOND",
  DigitalSecurityDetails = "DIGITAL_SECURITY_DETAILS",
  DigitalSecuritiesList = "DIGITAL_SECURITIES_LIST",
  DigitalSecurityMint = "DIGITAL_SECURITY_MINT",
  DigitalSecurityTransfer = "DIGITAL_SECURITY_TRANSFER",
  DigitalSecurityForceTransfer = "DIGITAL_SECURITY_FORCE_TRANSFER",
  DigitalSecurityRedeem = "DIGITAL_SECURITY_REDEEM",
  DigitalSecurityForceRedeem = "DIGITAL_SECURITY_FORCE_REDEEM",
}
