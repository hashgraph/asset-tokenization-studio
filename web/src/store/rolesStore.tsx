import { create } from "zustand";
import { devtools, persist } from "zustand/middleware";
import { SecurityRole } from "../utils/SecurityRole";

interface RolesStore {
  roles: SecurityRole[];
  setRoles: (roles: SecurityRole[]) => void;
}

const ROLES_STORE_KEY = "ROLES_STORE_KEY";

export const useRolesStore = create<RolesStore>()(
  persist(
    devtools((set) => ({
      roles: [],
      setRoles: (roles: SecurityRole[]) => set({ roles }),
    })),
    {
      name: ROLES_STORE_KEY,
    },
  ),
);
