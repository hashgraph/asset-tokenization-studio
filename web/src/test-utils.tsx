import React from "react";
import { ChakraProvider } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render, waitFor } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import { I18nextProvider } from "react-i18next";
import { Router } from "react-router-dom";
import i18n from "./i18n";
import theme from "./theme";
import userEvent from "@testing-library/user-event";

export const selectCalendar = async (
  component: RenderResult,
  id: string,
  day: string | number = new Date().getDate(),
) => {
  const calendar = component.getByTestId(id);
  await userEvent.click(calendar);

  await waitFor(() => {
    const daysToSelect = component.getAllByTestId(`day-${day}`);

    const dayToSelect = daysToSelect.find(
      (day) => !day.hasAttribute("disabled"),
    );

    if (dayToSelect) {
      userEvent.click(dayToSelect);
    }
  });

  await userEvent.click(document.body);
};

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: () => {},
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory,
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <I18nextProvider i18n={i18n}>
      <QueryClientProvider client={testQueryClient}>
        <ChakraProvider theme={theme}>
          <Router location={history.location} navigator={history}>
            {children}
          </Router>
        </ChakraProvider>
      </QueryClientProvider>
    </I18nextProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history,
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {},
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => (
      <AllProviders history={history}>{children}</AllProviders>
    ),
    ...options,
  });

export { customRender as render };
