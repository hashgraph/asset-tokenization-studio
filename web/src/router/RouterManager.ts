import type { ReverseParams } from "named-urls";
import { reverse } from "named-urls";
import { router } from ".";
import { RouteName } from "./RouteName";
import { RoutePath } from "./RoutePath";

export interface RouteParams {
  params?: ReverseParams;
  extra?: string;
  state?: object;
}

export class BaseRouterManager {
  constructor(private routes: Record<RouteName, string> = RoutePath) {}

  #getPath(name: RouteName, { params, extra = "" }: RouteParams) {
    const pattern = this.routes[name] + extra;
    return reverse(pattern, params);
  }

  to(name: RouteName, { state, ...params }: RouteParams = {}) {
    const pathname = this.#getPath(name, params);
    return router.navigate({ pathname }, { state });
  }

  getUrl(name: RouteName, params: RouteParams = {}) {
    const path = this.#getPath(name, params);
    return reverse(path);
  }

  goBack() {
    return router.navigate(-1);
  }

  goDashboard() {
    return this.to(RouteName.Dashboard);
  }
}

export const RouterManager = new BaseRouterManager(RoutePath);
