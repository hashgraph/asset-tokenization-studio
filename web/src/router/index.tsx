import { RouterProvider, createBrowserRouter } from "react-router-dom";
import { routes } from "./Routes";
import { useAccountStore } from "../store/accountStore";
import { useEffect } from "react";
import { useSecurityStore } from "../store/securityStore";

export const router = createBrowserRouter(routes);

const AppRouter = () => {
  const { isResetted, reset: resetAccountStore } = useAccountStore();
  const { reset: reserSecurityStore } = useSecurityStore();

  useEffect(() => {
    if (!isResetted) {
      resetAccountStore();
      reserSecurityStore();
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return <RouterProvider router={router} />;
};

export default AppRouter;
