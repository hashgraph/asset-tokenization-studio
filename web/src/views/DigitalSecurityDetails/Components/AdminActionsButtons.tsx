import { Link as RouterLink, useParams } from "react-router-dom";
import { useTranslation } from "react-i18next";
import { HStack, Menu, MenuButton, VStack } from "@chakra-ui/react";
import { Button } from "@hashgraph/uiComponents/Interaction";
import { RouteName } from "../../../router/RouteName";
import { RouterManager } from "../../../router/RouterManager";
import { PhosphorIcon, Toggle } from "@hashgraph/uiComponents";
import { DotsThreeVertical } from "@phosphor-icons/react";
import { Dropdown, DropdownItem } from "@hashgraph/uiComponents/DataDisplay";
import { Text } from "@hashgraph/uiComponents/Foundations";
import { usePauseSecurity } from "../../../hooks/queries/usePauseSecurity";
import { useUnpauseSecurity } from "../../../hooks/queries/useUnpauseSecurity";
import { PauseRequest } from "@hashgraph/securitytoken-sdk";
import { useGetIsPaused } from "../../../hooks/queries/useGetSecurityDetails";
import { useRolesStore } from "../../../store/rolesStore";
import { SecurityRole } from "../../../utils/SecurityRole";

export const AdminActionsButtons = () => {
  const { t: tButtons } = useTranslation("security", {
    keyPrefix: "details.actions",
  });
  const { id = "" } = useParams();
  const { data: isPaused, refetch } = useGetIsPaused(
    new PauseRequest({ securityId: id }),
  );
  const { roles } = useRolesStore();

  const { mutate: pauseSecurity, isLoading: isPauseLoading } = usePauseSecurity(
    { onSettled: () => refetch() },
  );
  const { mutate: unpauseSecurity, isLoading: isUnpauseLoading } =
    useUnpauseSecurity({ onSettled: () => refetch() });
  const handlePauseToggle = async () => {
    const pauseRequest = new PauseRequest({ securityId: id });
    if (isPaused) {
      unpauseSecurity(pauseRequest);
    } else {
      pauseSecurity(pauseRequest);
    }
  };

  const hasMinterRole = roles.find(
    (role) => role === SecurityRole._ISSUER_ROLE,
  );

  const hasControllerRole = roles.find(
    (role) => role === SecurityRole._CONTROLLER_ROLE,
  );

  const hasPauserRole = roles.find(
    (role) => role === SecurityRole._PAUSER_ROLE,
  );

  // TODO get from SDK the buttons to show depending of account's roles
  return (
    <HStack w="full" justifyContent="flex-end" gap={4}>
      {!isPaused && hasMinterRole && (
        <Button
          data-testid="mint-button"
          as={RouterLink}
          to={RouterManager.getUrl(RouteName.DigitalSecurityMint, {
            params: { id },
          })}
          variant="secondary"
        >
          {tButtons("mint")}
        </Button>
      )}
      {!isPaused && hasControllerRole && (
        <Button
          data-testid="force-transfer-button"
          as={RouterLink}
          to={RouterManager.getUrl(RouteName.DigitalSecurityForceTransfer, {
            params: { id },
          })}
          variant="secondary"
        >
          {tButtons("forceTransfer")}
        </Button>
      )}
      {(hasControllerRole || hasPauserRole) && (
        <Menu>
          <MenuButton
            data-testid="dropdown-actions-button"
            as={Button}
            size="sm"
            variant="tertiary"
            w={1}
            sx={{ minW: 0 }}
          >
            <PhosphorIcon as={DotsThreeVertical} />
          </MenuButton>
          <Dropdown
            w="146px"
            bg="neutral.dark.300"
            data-testid="dropdown-actions-menu"
          >
            {!isPaused && hasControllerRole ? (
              <DropdownItem
                label={tButtons("forceRedeem")}
                data-testid="force-redeem-button"
                as={RouterLink}
                onClick={() =>
                  RouterManager.to(RouteName.DigitalSecurityForceRedeem, {
                    params: { id },
                  })
                }
              />
            ) : (
              <></>
            )}
            {hasPauserRole ? (
              <VStack mt={2} w="full" align="flex-start" pb="2px">
                <HStack
                  w="full"
                  borderBottomColor="status.error.500"
                  borderBottomWidth="1px"
                  pb={1}
                >
                  <Text
                    textStyle="ElementsRegularXS"
                    color="status.error.500"
                    mt={2}
                  >
                    {tButtons("dangerZone.title")}
                  </Text>
                </HStack>
                <Text textStyle="ElementsRegularXS">
                  {tButtons("dangerZone.subtitle")}
                </Text>
                <HStack w="full" mt="10px">
                  <Toggle
                    data-testid="pauser-button"
                    label={tButtons(
                      `dangerZone.${
                        isPaused ? "buttonInactive" : "buttonActive"
                      }`,
                    )}
                    onChange={() => handlePauseToggle()}
                    isDisabled={isPauseLoading || isUnpauseLoading}
                  />
                </HStack>
              </VStack>
            ) : (
              <></>
            )}
          </Dropdown>
        </Menu>
      )}
    </HStack>
  );
};
