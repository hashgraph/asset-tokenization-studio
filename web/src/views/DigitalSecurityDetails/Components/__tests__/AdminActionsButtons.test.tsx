import { AdminActionsButtons } from "../AdminActionsButtons";
import { render } from "../../../../test-utils";
import { useRolesStore } from "../../../../store/rolesStore";
import { SecurityRole } from "../../../../utils/SecurityRole";
import userEvent from "@testing-library/user-event";
import { RouterManager } from "../../../../router/RouterManager";
import { RouteName } from "../../../../router/RouteName";
import { waitFor } from "@testing-library/react";
import { usePauseSecurity } from "../../../../hooks/queries/usePauseSecurity";
import { useUnpauseSecurity } from "../../../../hooks/queries/useUnpauseSecurity";

jest.mock("../../../../router/RouterManager", () => ({
  RouterManager: {
    ...jest.requireActual("../../../../router/RouterManager").RouterManager,
    getUrl: jest.fn(),
    to: jest.fn(),
  },
}));

jest.mock("../../../../hooks/queries/usePauseSecurity", () => ({
  usePauseSecurity: jest.fn(() => ({
    mutate: jest.fn(),
    isLoading: false,
  })),
}));

jest.mock("../../../../hooks/queries/useUnpauseSecurity", () => ({
  useUnpauseSecurity: jest.fn(() => ({
    mutate: jest.fn(),
    isLoading: false,
  })),
}));

const defaultAdminRole = [SecurityRole._DEFAULT_ADMIN_ROLE];
const initialStoreState = useRolesStore.getState();

describe(`${AdminActionsButtons.name}`, () => {
  beforeEach(() => {
    useRolesStore.setState(initialStoreState, true);
    jest.clearAllMocks();
  });

  const factoryComponent = () => render(<AdminActionsButtons />);

  test("should render correctly", () => {
    const component = factoryComponent();

    expect(component.asFragment()).toMatchSnapshot("defaultAdminRole");
  });

  test("by default admin has not minter role", () => {
    const component = factoryComponent();

    expect(component.queryByTestId("mint-button")).not.toBeInTheDocument();
  });

  test("by default admin has not controller role", () => {
    const component = factoryComponent();

    expect(
      component.queryByTestId("force-transfer-button"),
    ).not.toBeInTheDocument();
    expect(
      component.queryByTestId("force-redeem-button"),
    ).not.toBeInTheDocument();
  });

  test("by default admin has not pauser role", () => {
    const component = factoryComponent();

    expect(component.queryByTestId("pauser-button")).not.toBeInTheDocument();
  });

  describe("Admin has minter role", () => {
    beforeEach(() => {
      useRolesStore.setState({
        ...initialStoreState,
        roles: [...defaultAdminRole, SecurityRole._ISSUER_ROLE],
      });
    });

    test("should render correctly", () => {
      const component = factoryComponent();

      expect(component.asFragment()).toMatchSnapshot("minterRole");
    });

    test("should show mint button", () => {
      const component = factoryComponent();

      expect(component.getByTestId("mint-button")).toBeInTheDocument();
    });

    test("if click on mint button should redirect to mint page", async () => {
      const component = factoryComponent();

      const button = component.getByTestId("mint-button");
      expect(button).toBeInTheDocument();

      await userEvent.click(button);

      expect(RouterManager.getUrl).toHaveBeenCalled();
      expect(RouterManager.getUrl).toHaveBeenCalledWith(
        RouteName.DigitalSecurityMint,
        { params: { id: "" } },
      );
    });
  });

  describe("Admin has controller role", () => {
    beforeEach(() => {
      useRolesStore.setState({
        ...initialStoreState,
        roles: [...defaultAdminRole, SecurityRole._CONTROLLER_ROLE],
      });
    });

    test("should render correctly", () => {
      const component = factoryComponent();

      expect(component.asFragment()).toMatchSnapshot("controllerRole");
    });

    test("should show force transfer and force redeem buttons", () => {
      const component = factoryComponent();

      expect(
        component.getByTestId("force-transfer-button"),
      ).toBeInTheDocument();
      expect(component.getByTestId("force-redeem-button")).toBeInTheDocument();
    });

    test("if click on force transfer button should redirect to force transfer page", async () => {
      const component = factoryComponent();

      const button = component.getByTestId("force-transfer-button");
      expect(button).toBeInTheDocument();

      await userEvent.click(button);

      expect(RouterManager.getUrl).toHaveBeenCalled();
      expect(RouterManager.getUrl).toHaveBeenCalledWith(
        RouteName.DigitalSecurityForceTransfer,
        { params: { id: "" } },
      );
    });

    test("if click on force redeem button should redirect to force redeem page", async () => {
      const component = factoryComponent();

      const dropdown = component.getByTestId("dropdown-actions-button");
      expect(dropdown).toBeInTheDocument();

      await userEvent.click(dropdown);

      await waitFor(() => {
        expect(
          component.getByTestId("dropdown-actions-menu"),
        ).toBeInTheDocument();
      });

      const button = component.getByTestId("force-redeem-button");
      expect(button).toBeInTheDocument();

      await userEvent.click(button);

      expect(RouterManager.to).toHaveBeenCalled();
      expect(RouterManager.to).toHaveBeenCalledWith(
        RouteName.DigitalSecurityForceRedeem,
        { params: { id: "" } },
      );
    });
  });

  describe("Admin has pauser role", () => {
    beforeEach(() => {
      useRolesStore.setState({
        ...initialStoreState,
        roles: [...defaultAdminRole, SecurityRole._PAUSER_ROLE],
      });
    });

    test("should render correctly", () => {
      const component = factoryComponent();

      expect(component.asFragment()).toMatchSnapshot("pauserRole");
    });

    test("should show pauser toogle", () => {
      const component = factoryComponent();

      expect(component.getByTestId("pauser-button")).toBeInTheDocument();
    });

    test("if click on pauser button should call to pause or unpause security function", async () => {
      const component = factoryComponent();

      const dropdown = component.getByTestId("dropdown-actions-button");
      expect(dropdown).toBeInTheDocument();

      await userEvent.click(dropdown);

      await waitFor(() => {
        expect(
          component.getByTestId("dropdown-actions-menu"),
        ).toBeInTheDocument();
      });

      const button = component.getByTestId("pauser-button");
      expect(button).toBeInTheDocument();

      await userEvent.click(button);

      expect(usePauseSecurity).toHaveBeenCalled();
      expect(usePauseSecurity).toHaveBeenCalledWith({
        onSettled: expect.anything(),
      });

      await userEvent.click(button);

      expect(useUnpauseSecurity).toHaveBeenCalled();
      expect(useUnpauseSecurity).toHaveBeenCalledWith({
        onSettled: expect.anything(),
      });
    });
  });
});
