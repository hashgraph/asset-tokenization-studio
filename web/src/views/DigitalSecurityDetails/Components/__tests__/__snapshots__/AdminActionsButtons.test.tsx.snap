// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminActionsButtons Admin has controller role should render correctly: controllerRole 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1xa9bwv"
  >
    <a
      class="chakra-button css-vueqt9"
      data-testid="force-transfer-button"
      href="/"
    >
      Force transfer
    </a>
    <a
      class="chakra-button css-vueqt9"
      data-testid="force-redeem-button"
      href="/"
    >
      Force redeem
    </a>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`AdminActionsButtons Admin has minter role should render correctly: minterRole 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1xa9bwv"
  >
    <a
      class="chakra-button css-vueqt9"
      data-testid="mint-button"
      href="/"
    >
      Mint
    </a>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`AdminActionsButtons should render correctly: defaultAdminRole 1`] = `
<DocumentFragment>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
